python/ray/tune/suggest/basic_variant.py

```diff
@@ -11,6 +12,126 @@
+
+    def _load_value(self):
+        try:
+            self.next_value = next(self.iterable)
+        except StopIteration:
+            self._has_next = False
+
+    def has_next(self):
+        return self._has_next
+
+    def __next__(self):
+        if self.lazy_eval:
+            current_value = self.next_value
+            self._load_value()
+            return current_value
+        current_value = self.iterable.pop(0)
+        self._has_next = bool(self.iterable)
+        return current_value
+
+
+class _TrialIterator:
+    def __init__(self,
+                 uuid_prefix,
+                 num_samples,
+                 unresolved_spec,
+                 output_path="",
+                 points_to_evaluate=None,
+                 lazy_eval=False,
+                 start=0):
+        self.parser = make_parser()
+        self.num_samples = num_samples
+        self.uuid_prefix = uuid_prefix
+        self.num_samples_left = num_samples
+        self.unresolved_spec = unresolved_spec
+        self.output_path = output_path
+        self.points_to_evaluate = points_to_evaluate or []
+        self.num_points_to_evaluate = len(self.points_to_evaluate)
+        self.counter = start
+        self.lazy_eval = lazy_eval
+        self.variants = None
+
+    def set_variants_from_preset(self, config):
+        # This can cause OOM if not careful... perhaps we can have a
+        # check here?
+        self.variants = _VariantIterator(
+            get_preset_variants(self.unresolved_spec, config),
+            lazy_eval=self.lazy_eval)
+
+    def set_generated_variants(self):
+        self.variants = _VariantIterator(
+            generate_variants(self.unresolved_spec), lazy_eval=self.lazy_eval)
+
+    def create_trial(self, resolved_vars, spec):
+        trial_id = self.uuid_prefix + ("%05d" % self.counter)
+        experiment_tag = str(self.counter)
+        # Always append resolved vars to experiment tag?
+        if resolved_vars:
+            experiment_tag += "_{}".format(format_vars(resolved_vars))
+        self.counter += 1
+        return create_trial_from_spec(
+            spec,
+            self.output_path,
+            self.parser,
+            evaluated_params=flatten_resolved_vars(resolved_vars),
+            trial_id=trial_id,
+            experiment_tag=experiment_tag)
+
+    def __next__(self):
+        """Generates Trial objects with the variant generation process.
+
+        Uses a fixed point iteration to resolve variants. All trials
+        should be able to be generated at once.
+
+        See also: `ray.tune.suggest.variant_generator`.
+
+        Returns:
+            Trial object
+        """
+
+        if "run" not in self.unresolved_spec:
+            raise TuneError("Must specify `run` in {}".format(
+                self.unresolved_spec))
+
+        if self.variants and self.variants.has_next():
+            resolved_vars, spec = next(self.variants)
+            return self.create_trial(resolved_vars, spec)
```

Comment:
This was a little bit tricky to find out. Maybe add a comment here to say that this will be skipped on the first call of `__next__()` and the next block initializes/updates `self.variants`